binder/binder.cpp

```diff
@@ -0,0 +1,1150 @@
+
+    if (!g_strcmp0(iface, GNSS_CALLBACK)) {
+        GBinderReader reader;
+
+        gbinder_remote_request_init_reader(req, &reader);
+        switch (code) {
+        case GNSS_LOCATION_CB:
+            {
+            Location loc;
+
+            const GnssLocation *location = geoclue_binder_gnss_decode_struct
+                (GnssLocation, &reader);
+
+            loc.setTimestamp(location->timestamp);
+
+            if (location->gnssLocationFlags & HYBRIS_GNSS_LOCATION_HAS_LAT_LONG) {
+                loc.setLatitude(location->latitudeDegrees);
+                loc.setLongitude(location->longitudeDegrees);
+            }
+
+            if (location->gnssLocationFlags & HYBRIS_GNSS_LOCATION_HAS_ALTITUDE)
+                loc.setAltitude(location->altitudeMeters);
+
+            if (location->gnssLocationFlags & HYBRIS_GNSS_LOCATION_HAS_SPEED)
+                loc.setSpeed(location->speedMetersPerSec);
+
+            if (location->gnssLocationFlags & HYBRIS_GNSS_LOCATION_HAS_BEARING)
+                loc.setDirection(location->bearingDegrees);
+
+            if ((location->gnssLocationFlags & HYBRIS_GNSS_LOCATION_HAS_HORIZONTAL_ACCURACY) ||
+                (location->gnssLocationFlags & HYBRIS_GNSS_LOCATION_HAS_VERTICAL_ACCURACY)) {
+                Accuracy accuracy;
+                if (location->gnssLocationFlags & HYBRIS_GNSS_LOCATION_HAS_HORIZONTAL_ACCURACY) {
+                    accuracy.setHorizontal(location->horizontalAccuracyMeters);
+                }
+                if (location->gnssLocationFlags & HYBRIS_GNSS_LOCATION_HAS_VERTICAL_ACCURACY) {
+                    accuracy.setVertical(location->verticalAccuracyMeters);
+                }
+                loc.setAccuracy(accuracy);
+            }
+
+            QMetaObject::invokeMethod(staticProvider, "setLocation", Qt::QueuedConnection,
+                                      Q_ARG(Location, loc));
+            }
+            break;
+        case GNSS_STATUS_CB:
+            {
+            guint32 stat;
+            if (gbinder_reader_read_uint32(&reader, &stat)) {
+                if (stat == HYBRIS_GNSS_STATUS_ENGINE_ON) {
+                    QMetaObject::invokeMethod(staticProvider, "engineOn", Qt::QueuedConnection);
+                }
+                if (stat == HYBRIS_GNSS_STATUS_ENGINE_OFF) {
+                    QMetaObject::invokeMethod(staticProvider, "engineOff", Qt::QueuedConnection);
+                }
+            }
+            }
+            break;
+        case GNSS_SV_STATUS_CB:
+            {
+            const GnssSvStatus* svStatus = geoclue_binder_gnss_decode_struct
+                (GnssSvStatus, &reader);
+
+            QList<SatelliteInfo> satellites;
+            QList<int> usedPrns;
+
+            for (int i = 0; i < svStatus->numSvs; ++i) {
+                SatelliteInfo satInfo;
+                GnssSvInfo svInfo = svStatus->gnssSvList.data[i];
+                satInfo.setPrn(svInfo.svid);
+                satInfo.setSnr(svInfo.cN0Dbhz);
```

Comment:
I guess the signal to noise ratio is usually a value from 0.0 to 1.0, but the carrier/noise value may not be?  Should this be normalised somehow to a valid snr value?  (I checked https://doc-snapshots.qt.io/qt5-5.11/qgeosatelliteinfo.html#signalStrength and it seems that this isn't documented to be SNR specifically, and in fact has no documented semantics at all, which is unfortunate.  But value-range consistency would be appreciated nonetheless.)

Reply:
I will have to check how the satellite info in handled in other components and decide how to handle that.

Reply:
I checked the values we get from the snr field in older adaptations and I saw on FP2 at least values between 0-40 so I think we can still use the raw value we get from the cN0Dbhz field of GnssSvInfo.